packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts

```diff
@@ -1156,15 +1159,92 @@ export class ExchangeWrapper extends ContractWrapper {
         const balanceAllowanceStore = new BalanceAndProxyAllowanceLazyStore(balanceAllowanceFetcher);
         const exchangeTradeSimulator = new ExchangeTransferSimulator(balanceAllowanceStore);
 
-        const expectedFillTakerTokenAmountIfExists = opts.expectedFillTakerTokenAmount;
         const filledCancelledFetcher = new OrderFilledCancelledFetcher(this, BlockParamLiteral.Latest);
+
+        let fillableTakerAssetAmount;
+        const shouldValidateRemainingOrderAmountIsFillable = _.isUndefined(opts.validateRemainingOrderAmountIsFillable)
+            ? true
+            : opts.validateRemainingOrderAmountIsFillable;
+        const filledTakerTokenAmount = await this.getFilledTakerAssetAmountAsync(
+            orderHashUtils.getOrderHashHex(signedOrder),
+        );
+        if (opts.expectedFillTakerTokenAmount) {
+            // If the caller has specified a taker fill amount, we use this for all validation
+            fillableTakerAssetAmount = opts.expectedFillTakerTokenAmount;
+        } else if (shouldValidateRemainingOrderAmountIsFillable) {
+            // Historically if a fill amount was not specified we would default to the amount
+            // left on the order.
+            fillableTakerAssetAmount = signedOrder.takerAssetAmount.minus(filledTakerTokenAmount);
+        } else {
+            const makerAssetBalance = await balanceAllowanceStore.getBalanceAsync(
```

Comment:
Let's move this logic into an `order_utils` method so that others can compute this for other purposes (e.g Coordinator server needs it too).

Reply:
refactored to use `orderStateUtils.getOpenOrderRelevantStateAsync`